OSCORE: Add support for more AES CCM algorithms

rikard-sics opened the pull request:
Added support for using the following algorithms with OSCORE:
AES-CCM-64-64-128
AES-CCM-16-128-128
AES-CCM-64-128-128

Previously only AES-CCM-16-64-128 was supported.

The actual encryption/decryption uses the CCMBlockCipher.encrypt and CCMBlockCipher.decrypt methods from Scandium.

boaks commented on cf-oscore/src/main/java/org/eclipse/californium/oscore/OSCoreCtx.java:
```diff
@@ -834,7 +842,11 @@ public synchronized void rollBack() {
 	private static int ivLength(AlgorithmID alg) {
```
I would prefer, if this is moved to "EncryptCommon" (use the finals there) and made public.

rikard-sics replied:
Yes, that makes sense. This has now been done in commit 8c0bee8df9e1eb0fb862725a36d47c2ed6f4965a

boaks commented on cf-oscore/src/main/java/org/eclipse/californium/oscore/OSCoreCtx.java:
```diff
@@ -689,14 +689,22 @@ public void setSeqMax(int seqMax) {
-				id_length = 7;
-				key_length = common_alg.getKeySize() / 8; // 16;
-			} else {
+				id_length = iv_length - 6; // RFC section 5.2
```
If the iv-length is 7, the resulting id-length is 1. Is that intended?

rikard-sics replied:
Yes, that is correct. That is why OSCORE only supports AEAD algorithms with nonces 7 bytes or more.

Having a maximum length of the sender/recipient ID be 1 byte only is small, but there is also the ID Context that can be used to identify a specific context (if more than 256 are needed).

boaks commented on cf-oscore/src/main/java/org/eclipse/californium/cose/EncryptCommon.java:
```diff
@@ -97,7 +104,12 @@ void encryptWithKey(byte[] rgbKey) throws CoseException, IllegalStateException {
 	private int getAES_CCM_IVSize(AlgorithmID alg) throws CoseException {
```
Now the difference to the new `public static int ivLength(AlgorithmID alg)` seems to be very small. Maybe you redesign it a little to remove this one?

rikard-sics replied:
True, they are very similar now. I will redesign things to only keep 1.

Just to clarify, you mean removing getAES_CCM_IVSize and keeping ivLength, or the opposite?

boaks replied:
I guess, the `public static` one is the one to keep. if you keep the -1 or the exception depends on the calling contexts. Just check, what will fit better for you.

boaks commented on cf-oscore/src/main/java/org/eclipse/californium/oscore/OSCoreCtx.java:
```diff
@@ -689,14 +690,22 @@ public void setSeqMax(int seqMax) {
+			case AES_CCM_64_64_128:
+			case AES_CCM_16_128_128:
+			case AES_CCM_64_128_128:
+				iv_length = EncryptCommon.ivLength(common_alg);
```
`ivLength` would return a `-1` for not supported algorithms (or maybe a exception after redesign). Checking for the returned value (or exception) would obsolete the repeated "switch(algo)".

boaks replied:
It's OK, if you want to keep this switch. I would replace it with `iv_length > 0`.

rikard-sics replied:
I will update this one also. I was too focused on the EncryptCommon and missed this one.

boaks commented on cf-oscore/src/main/java/org/eclipse/californium/cose/EncryptCommon.java:
```diff
@@ -218,4 +232,20 @@ private void AES_CCM_Encrypt(AlgorithmID alg, byte[] rgbKey) throws CoseExceptio
+		default:
+			return -1;
+		}
+	}
```
Maybe adding a  `boolean isSupportedAesCcm(AlgorithmID alg)` obsoletes some general `switch(alg)` as above in line 69 or 91. Or use return value of this `ivLength()` instead of the `switch(alg)`.

rikard-sics commented:
I made a new commit now baed0fb9426cbc57dd303a290de9e474f8cb762c with some updates from our discussion.

Changes done:

> Removed the duplicate method for getting the IV length
> Added a method for checking if an algorithm is supported
> Removed some switch statements

boaks commented:
You may replace the switch (or not :-). If you ready, please rebase and squash.

rikard-sics commented:
> You may replace the switch (or not :-). If you ready, please rebase and squash.

I updated removed that switch statement and have now rebased and squashed.
